lib/ai_bot/playground.rb

```diff
@@ -3,26 +3,101 @@
 
+      def self.schedule_reply(post)
+        bot_ids = DiscourseAi::AiBot::EntryPoint::BOT_USER_IDS
+
+        return if bot_ids.include?(post.user_id)
+        if AiPersona.mentionables.any? { |mentionable| mentionable[:user_id] == post.user_id }
+          return
+        end
+
+        bot_user = nil
+        mentioned = nil
+
+        if post.topic.private_message?
+          bot_user = post.topic.topic_allowed_users.where(user_id: bot_ids).first&.user
+        end
+
+        if AiPersona.mentionables.length > 0
+          mentions = post.mentions.map(&:downcase)
+          mentioned =
+            AiPersona.mentionables.find do |mentionable|
+              mentions.include?(mentionable[:username]) &&
+                (post.user.group_ids & mentionable[:allowed_group_ids]).present?
+            end
+
+          # PM always takes precedence
+          if mentioned && !bot_user
+            model_without_provider = mentioned[:default_llm].split(":").last
+            user_id =
+              DiscourseAi::AiBot::EntryPoint.map_bot_model_to_user_id(model_without_provider)
+
+            if !user_id
+              Rails.logger.warn(
+                "Model #{mentioned[:default_llm]} not found for persona #{mentioned[:username]}",
+              )
+              if Rails.env.development? || Rails.env.test?
+                raise "Model #{mentioned[:default_llm]} not found for persona #{mentioned[:username]}"
+              end
+            else
+              bot_user = User.find_by(id: user_id)
+            end
+          end
+        end
+
+        if bot_user
+          persona_id = mentioned&.dig(:id) || post.topic.custom_fields["ai_persona_id"]
+          persona = nil
+
+          if persona_id
+            persona =
+              DiscourseAi::AiBot::Personas::Persona.find_by(user: post.user, id: persona_id.to_i)
+          end
+
+          if !persona && persona_name = post.topic.custom_fields["ai_persona"]
+            persona =
+              DiscourseAi::AiBot::Personas::Persona.find_by(user: post.user, name: persona_name)
+          end
+
+          persona ||= DiscourseAi::AiBot::Personas::General
+
+          bot = DiscourseAi::AiBot::Bot.as(bot_user, persona: persona.new)
+          new(bot).update_playground_with(post)
+        end
+      end
+
       def initialize(bot)
         @bot = bot
       end
 
       def update_playground_with(post)
-        if can_attach?(post) && bot.bot_user
-          schedule_playground_titling(post, bot.bot_user)
-          schedule_bot_reply(post, bot.bot_user)
+        if can_attach?(post)
+          schedule_playground_titling(post)
+          schedule_bot_reply(post)
         end
       end
 
       def conversation_context(post)
         # Pay attention to the `post_number <= ?` here.
         # We want to inject the last post as context because they are translated differently.
+
+        # also setting default to 40, allowing huge contexts costs lots of tokens
+        max_posts = 40
```

Comment:
```suggestion
        max_posts = DEFAULT_MAX_CONTEXT_POSTS
```

Probably should have a const for this